runtime_builders/publish_runtime.py

```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/python
+
+# Copyright 2016 Google Inc. All rights reserved.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+from datetime import datetime
+import json
+import logging
+import os
+import regex
+from ruamel import yaml
+import subprocess
+import sys
+import tempfile
+
+from google.cloud import storage
+
+LANGUAGES = [
+    'java',
+    'python',
+    'ruby',
+    'nodejs',
+    'golang',
+    'dotnet',
+    'php'
+]
+
+TAG_REGEX = '(?<=:|@)(.*)'
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--infile', '-i',
+                        help='templated cloudbuild config file.',
+                        required=True)
+    parser.add_argument('--bucket', '-b',
+                        help='GCS bucket to publish runtime to',
+                        default='runtime-builders')
+    parser.add_argument('--language',
+                        help='the language associated with this builder',
+                        required=True,
+                        choices=LANGUAGES)
+    args = parser.parse_args()
+
+    if args.language not in LANGUAGES:
+        logging.error('Invalid language \'{0}\' specified! '
+                      'Options:'.format(args.language))
+        logging.error(LANGUAGES)
+        return 1
+
+    templated_file = _resolve_tags(args.infile)
+    _publish_to_gcs(templated_file, args.language, args.bucket)
+
+
+def _resolve_tags(config_file):
+    """
+    Given a templated YAML cloudbuild config file, parse it, resolve image tags
+    on each build step's image to the corresponding digest, and write new
+    config with fully qualified images to temporary file for upload to GCS.
+
+    Keyword arguments:
+    config_file -- string representing path to
+    templated cloudbuild YAML config file
+
+    Return value:
+    path to temporary file containing fully qualified config file, to be
+    published to GCS.
+    """
+    with open(config_file, 'r') as infile:
+        try:
+            config = yaml.round_trip_load(infile)
+
+            for step in config.get('steps'):
+                image = step.get('name')
+                templated_step = _resolve_tag(image)
+                step['name'] = templated_step
+
+            s = yaml.round_trip_dump(config)
+
+            fd, ofile = tempfile.mkstemp()
+            with open(ofile, 'w') as outfile:
+                outfile.write(s)
```

Comment:
Hmm, do we need to actually write this to a file? Could we keep it in memory and upload it that way?